File: 01-Programming-Languages/Go/OOP/oop.go
// Package main demonstrates Go's approach to object-oriented programming
// using interfaces, struct embedding, and composition instead of classes and inheritance.
package main

import (
	"fmt"
	"math"
)

// Shape interface defines the contract for all shapes.
// In Go, interfaces are satisfied implicitly - no "implements" keyword needed.
type Shape interface {
	Area() float64
	Perimeter() float64
	String() string
}

// Circle represents a circle shape.
type Circle struct {
	Radius float64
}

// Area calculates the area of the circle.
// Time complexity: O(1)
func (c Circle) Area() float64 {
	return math.Pi * c.Radius * c.Radius
}

// Perimeter calculates the circumference of the circle.
// Time complexity: O(1)
func (c Circle) Perimeter() float64 {
	return 2 * math.Pi * c.Radius
}

// String returns a string representation of the circle.
func (c Circle) String() string {
	return fmt.Sprintf("Circle{Radius: %.2f}", c.Radius)
}

// Rectangle represents a rectangle shape.
type Rectangle struct {
	Width, Height float64
}

// Area calculates the area of the rectangle.
// Time complexity: O(1)
func (r Rectangle) Area() float64 {
	return r.Width * r.Height
}

// Perimeter calculates the perimeter of the rectangle.
// Time complexity: O(1)
func (r Rectangle) Perimeter() float64 {
	return 2 * (r.Width + r.Height)
}

// String returns a string representation of the rectangle.
func (r Rectangle) String() string {
	return fmt.Sprintf("Rectangle{Width: %.2f, Height: %.2f}", r.Width, r.Height)
}

// Triangle represents a triangle shape using three sides.
type Triangle struct {
	A, B, C float64 // Sides
}

// Area calculates the area using Heron's formula.
// Time complexity: O(1)
func (t Triangle) Area() float64 {
	s := (t.A + t.B + t.C) / 2
	return math.Sqrt(s * (s - t.A) * (s - t.B) * (s - t.C))
}

// Perimeter calculates the perimeter of the triangle.
// Time complexity: O(1)
func (t Triangle) Perimeter() float64 {
	return t.A + t.B + t.C
}

// String returns a string representation of the triangle.
func (t Triangle) String() string {
	return fmt.Sprintf("Triangle{A: %.2f, B: %.2f, C: %.2f}", t.A, t.B, t.C)
}

// Square embeds Rectangle to reuse its functionality.
// This demonstrates composition over inheritance.
type Square struct {
	Rectangle // Embedded struct
}

// NewSquare creates a new square with the given side length.
func NewSquare(side float64) Square {
	return Square{
		Rectangle: Rectangle{Width: side, Height: side},
	}
}

// Area is promoted from Rectangle, but we can override it if needed.
// Time complexity: O(1)
func (s Square) Area() float64 {
	return s.Rectangle.Area() // Explicitly call embedded method
}

// String returns a string representation of the square.
func (s Square) String() string {
	return fmt.Sprintf("Square{Side: %.2f}", s.Rectangle.Width)
}

// ShapeCollection demonstrates working with a collection of shapes.
type ShapeCollection struct {
	shapes []Shape
}

// NewShapeCollection creates a new shape collection.
func NewShapeCollection() *ShapeCollection {
	return &ShapeCollection{
		shapes: make([]Shape, 0),
	}
}

// Add adds a shape to the collection.
func (sc *ShapeCollection) Add(s Shape) {
	sc.shapes = append(sc.shapes, s)
}

// TotalArea calculates the sum of areas of all shapes.
// Time complexity: O(n) where n is the number of shapes
func (sc *ShapeCollection) TotalArea() float64 {
	total := 0.0
	for _, shape := range sc.shapes {
		total += shape.Area()
	}
	return total
}

// TotalPerimeter calculates the sum of perimeters of all shapes.
// Time complexity: O(n) where n is the number of shapes
func (sc *ShapeCollection) TotalPerimeter() float64 {
	total := 0.0
	for _, shape := range sc.shapes {
		total += shape.Perimeter()
	}
	return total
}

// LargestShape returns the shape with the largest area.
// Time complexity: O(n)
func (sc *ShapeCollection) LargestShape() Shape {
	if len(sc.shapes) == 0 {
		return nil
	}

	largest := sc.shapes[0]
	for _, shape := range sc.shapes[1:] {
		if shape.Area() > largest.Area() {
			largest = shape
		}
	}
	return largest
}

// PrintAll prints information about all shapes.
func (sc *ShapeCollection) PrintAll() {
	fmt.Println("Shape Collection:")
	fmt.Println("================")
	for i, shape := range sc.shapes {
		fmt.Printf("%d. %s\n", i+1, shape)
		fmt.Printf("   Area: %.2f, Perimeter: %.2f\n", shape.Area(), shape.Perimeter())
	}
	fmt.Printf("\nTotal Area: %.2f\n", sc.TotalArea())
	fmt.Printf("Total Perimeter: %.2f\n", sc.TotalPerimeter())
	if largest := sc.LargestShape(); largest != nil {
		fmt.Printf("Largest Shape: %s (Area: %.2f)\n", largest, largest.Area())
	}
}

// Matrix demonstrates struct embedding for code reuse.
type Matrix struct {
	Rows, Cols int
	data       [][]float64
}

// NewMatrix creates a new matrix with the given dimensions.
func NewMatrix(rows, cols int) *Matrix {
	data := make([][]float64, rows)
	for i := range data {
		data[i] = make([]float64, cols)
	}
	return &Matrix{Rows: rows, Cols: cols, data: data}
}

// Set sets a value in the matrix.
func (m *Matrix) Set(row, col int, value float64) {
	if row >= 0 && row < m.Rows && col >= 0 && col < m.Cols {
		m.data[row][col] = value
	}
}

// Get gets a value from the matrix.
func (m *Matrix) Get(row, col int) float64 {
	if row >= 0 && row < m.Rows && col >= 0 && col < m.Cols {
		return m.data[row][col]
	}
	return 0
}

// String returns a string representation of the matrix.
func (m *Matrix) String() string {
	result := fmt.Sprintf("Matrix{%dx%d}:\n", m.Rows, m.Cols)
	for i := 0; i < m.Rows; i++ {
		for j := 0; j < m.Cols; j++ {
			result += fmt.Sprintf("%6.2f ", m.data[i][j])
		}
		result += "\n"
	}
	return result
}

func main() {
	fmt.Println("=== Go's Approach to Object-Oriented Programming ===")
	fmt.Println("Demonstrating interfaces, struct embedding, and composition")
	fmt.Println()

	// Create different shapes
	circle := Circle{Radius: 5.0}
	rectangle := Rectangle{Width: 4.0, Height: 6.0}
	triangle := Triangle{A: 3.0, B: 4.0, C: 5.0}
	square := NewSquare(4.0)

	// Demonstrate that all shapes implement the Shape interface
	fmt.Println("Individual Shapes:")
	fmt.Printf("%s\n", circle)
	fmt.Printf("  Area: %.2f, Perimeter: %.2f\n\n", circle.Area(), circle.Perimeter())

	fmt.Printf("%s\n", rectangle)
	fmt.Printf("  Area: %.2f, Perimeter: %.2f\n\n", rectangle.Area(), rectangle.Perimeter())

	fmt.Printf("%s\n", triangle)
	fmt.Printf("  Area: %.2f, Perimeter: %.2f\n\n", triangle.Area(), triangle.Perimeter())

	fmt.Printf("%s\n", square)
	fmt.Printf("  Area: %.2f, Perimeter: %.2f\n\n", square.Area(), square.Perimeter())

	// Use shapes polymorphically through the Shape interface
	fmt.Println("Polymorphic usage:")
	shapes := []Shape{circle, rectangle, triangle, square}
	for _, shape := range shapes {
		fmt.Printf("Shape: %s\n", shape)
	}

	// Use ShapeCollection to group shapes
	fmt.Println("\n=== Shape Collection Demo ===")
	collection := NewShapeCollection()
	collection.Add(circle)
	collection.Add(rectangle)
	collection.Add(triangle)
	collection.Add(square)
	collection.PrintAll()

	// Demonstrate method promotion through embedding
	fmt.Println("\n=== Struct Embedding Demo ===")
	fmt.Println("Square inherits from Rectangle through embedding:")
	fmt.Printf("Square area (promoted): %.2f\n", square.Area())
	fmt.Printf("Square perimeter (promoted): %.2f\n", square.Perimeter())

	// Matrix example showing code reuse
	fmt.Println("\n=== Matrix with Embedded Structs ===")
	m := NewMatrix(2, 2)
	m.Set(0, 0, 1.0)
	m.Set(0, 1, 2.0)
	m.Set(1, 0, 3.0)
	m.Set(1, 1, 4.0)
	fmt.Println(m)

	// Interface satisfaction check
	fmt.Println("=== Interface Satisfaction ===")
	fmt.Printf("Circle satisfies Shape: %v\n", doesImplement(circle))
	fmt.Printf("Rectangle satisfies Shape: %v\n", doesImplement(rectangle))
	fmt.Printf("int satisfies Shape: %v\n", doesImplement(42))
}

// doesImplement checks if a type implements the Shape interface.
// This is a runtime check; in Go, interface satisfaction is usually verified at compile time.
func doesImplement(shape interface{}) bool {
	_, ok := shape.(Shape)
	return ok
}

File: 01-Programming-Languages/Rust/OOP/oop.rs
// This file demonstrates Rust's approach to object-oriented programming
// using traits for polymorphism instead of classes and inheritance.

use std::f64::consts::PI;

// Define a trait for shapes (similar to interface in other languages)
trait Shape {
    fn area(&self) -> f64;
    fn perimeter(&self) -> f64;
    fn description(&self) -> String;
    
    // Default implementation
    fn is_congruent(&self, other: &dyn Shape) -> bool {
        (self.area() - other.area()).abs() < f64::EPSILON
    }
}

// Circle struct
#[derive(Debug, Clone)]
struct Circle {
    radius: f64,
}

impl Circle {
    // Associated function (constructor)
    fn new(radius: f64) -> Self {
        Circle { radius }
    }
    
    // Method specific to Circle
    fn diameter(&self) -> f64 {
        2.0 * self.radius
    }
}

// Implement Shape trait for Circle
impl Shape for Circle {
    fn area(&self) -> f64 {
        PI * self.radius * self.radius
    }
    
    fn perimeter(&self) -> f64 {
        2.0 * PI * self.radius
    }
    
    fn description(&self) -> String {
        format!("Circle with radius {:.2}", self.radius)
    }
    
    fn is_congruent(&self, other: &dyn Shape) -> bool {
        if let Some(other_circle) = other.downcast_ref::<Circle>() {
            (self.radius - other_circle.radius).abs() < f64::EPSILON
        } else {
            false
        }
    }
}

// Rectangle struct
#[derive(Debug, Clone)]
struct Rectangle {
    width: f64,
    height: f64,
}

impl Rectangle {
    fn new(width: f64, height: f64) -> Self {
        Rectangle { width, height }
    }
    
    fn is_square(&self) -> bool {
        (self.width - self.height).abs() < f64::EPSILON
    }
}

impl Shape for Rectangle {
    fn area(&self) -> f64 {
        self.width * self.height
    }
    
    fn perimeter(&self) -> f64 {
        2.0 * (self.width + self.height)
    }
    
    fn description(&self) -> String {
        format!("Rectangle with width {:.2} and height {:.2}", self.width, self.height)
    }
    
    fn is_congruent(&self, other: &dyn Shape) -> bool {
        if let Some(other_rect) = other.downcast_ref::<Rectangle>() {
            ((self.width - other_rect.width).abs() < f64::EPSILON
                && (self.height - other_rect.height).abs() < f64::EPSILON)
                || ((self.width - other_rect.height).abs() < f64::EPSILON
                    && (self.height - other_rect.width).abs() < f64::EPSILON)
        } else {
            false
        }
    }
}

// Triangle struct (three sides)
#[derive(Debug, Clone)]
struct Triangle {
    a: f64,
    b: f64,
    c: f64,
}

impl Triangle {
    fn new(a: f64, b: f64, c: f64) -> Result<Self, &'static str> {
        // Validate triangle inequality
        if a + b <= c || a + c <= b || b + c <= a {
            return Err("Invalid triangle: violates triangle inequality");
        }
        Ok(Triangle { a, b, c })
    }
    
    fn is_equilateral(&self) -> bool {
        (self.a - self.b).abs() < f64::EPSILON
            && (self.b - self.c).abs() < f64::EPSILON
    }
    
    fn is_isosceles(&self) -> bool {
        (self.a - self.b).abs() < f64::EPSILON
            || (self.a - self.c).abs() < f64::EPSILON
            || (self.b - self.c).abs() < f64::EPSILON
    }
}

impl Shape for Triangle {
    fn area(&self) -> f64 {
        // Heron's formula
        let s = (self.a + self.b + self.c) / 2.0;
        (s * (s - self.a) * (s - self.b) * (s - self.c)).sqrt()
    }
    
    fn perimeter(&self) -> f64 {
        self.a + self.b + self.c
    }
    
    fn description(&self) -> String {
        format!(
            "Triangle with sides {:.2}, {:.2}, {:.2}",
            self.a, self.b, self.c
        )
    }
}

// Square struct (special case of Rectangle)
#[derive(Debug, Clone)]
struct Square {
    side: f64,
}

impl Square {
    fn new(side: f64) -> Self {
        Square { side }
    }
}

impl Shape for Square {
    fn area(&self) -> f64 {
        self.side * self.side
    }
    
    fn perimeter(&self) -> f64 {
        4.0 * self.side
    }
    
    fn description(&self) -> String {
        format!("Square with side {:.2}", self.side)
    }
    
    fn is_congruent(&self, other: &dyn Shape) -> bool {
        if let Some(other_square) = other.downcast_ref::<Square>() {
            (self.side - other_square.side).abs() < f64::EPSILON
        } else {
            false
        }
    }
}

// Implement Display trait for all shapes
impl std::fmt::Display for dyn Shape {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{}", self.description())
    }
}

// Collection of shapes
struct ShapeCollection {
    shapes: Vec<Box<dyn Shape>>,
}

impl ShapeCollection {
    fn new() -> Self {
        ShapeCollection {
            shapes: Vec::new(),
        }
    }
    
    fn add(&mut self, shape: Box<dyn Shape>) {
        self.shapes.push(shape);
    }
    
    fn total_area(&self) -> f64 {
        self.shapes.iter().map(|s| s.area()).sum()
    }
    
    fn total_perimeter(&self) -> f64 {
        self.shapes.iter().map(|s| s.perimeter()).sum()
    }
    
    fn largest_shape(&self) -> Option<&dyn Shape> {
        self.shapes
            .iter()
            .max_by(|a, b| a.area().partial_cmp(&b.area()).unwrap())
            .map(|s| s.as_ref())
    }
    
    fn print_all(&self) {
        println!("Shape Collection:");
        println!("================");
        for (i, shape) in self.shapes.iter().enumerate() {
            println!("{}. {}", i + 1, shape);
            println!("   Area: {:.2}, Perimeter: {:.2}", shape.area(), shape.perimeter());
        }
        println!();
        println!("Total Area: {:.2}", self.total_area());
        println!("Total Perimeter: {:.2}", self.total_perimeter());
        if let Some(largest) = self.largest_shape() {
            println!("Largest Shape: {} (Area: {:.2})", largest, largest.area());
        }
    }
}

// Trait for shapes that can be scaled
trait Scalable {
    fn scale(&mut self, factor: f64);
}

// Implement Scalable for different shapes
impl Scalable for Circle {
    fn scale(&mut self, factor: f64) {
        self.radius *= factor;
    }
}

impl Scalable for Rectangle {
    fn scale(&mut self, factor: f64) {
        self.width *= factor;
        self.height *= factor;
    }
}

impl Scalable for Square {
    fn scale(&mut self, factor: f64) {
        self.side *= factor;
    }
}

// Implement Scalable for Triangle
impl Scalable for Triangle {
    fn scale(&mut self, factor: f64) {
        self.a *= factor;
        self.b *= factor;
        self.c *= factor;
    }
}

// Demonstrate trait objects and dynamic dispatch
fn print_shape_info(shape: &dyn Shape) {
    println!("Shape: {}", shape);
    println!("  Area: {:.2}", shape.area());
    println!("  Perimeter: {:.2}", shape.perimeter());
}

// Demonstrate generic function with trait bounds
fn compare_shapes<T: Shape, U: Shape>(shape1: &T, shape2: &U) {
    println!("Comparing shapes:");
    println!("  {} vs {}", shape1.description(), shape2.description());
    println!("  Same area: {}", (shape1.area() - shape2.area()).abs() < f64::EPSILON);
    println!("  Same perimeter: {}", (shape1.perimeter() - shape2.perimeter()).abs() < f64::EPSILON);
}

fn main() {
    println!("=== Rust's Approach to Object-Oriented Programming ===");
    println!("Demonstrating traits, enums, and composition instead of classes and inheritance");
    println!();
    
    // Create different shapes
    let circle = Circle::new(5.0);
    let rectangle = Rectangle::new(4.0, 6.0);
    let triangle = Triangle::new(3.0, 4.0, 5.0).unwrap();
    let square = Square::new(4.0);
    
    // Demonstrate trait implementations
    println!("Individual Shapes:");
    println!("{}", circle);
    println!("  Area: {:.2}, Perimeter: {:.2}", circle.area(), circle.perimeter());
    println!();
    
    println!("{}", rectangle);
    println!("  Area: {:.2}, Perimeter: {:.2}", rectangle.area(), rectangle.perimeter());
    println!();
    
    println!("{}", triangle);
    println!("  Area: {:.2}, Perimeter: {:.2}", triangle.area(), triangle.perimeter());
    println!();
    
    println!("{}", square);
    println!("  Area: {:.2}, Perimeter: {:.2}", square.area(), square.perimeter());
    println!();
    
    // Demonstrate trait object usage (dynamic dispatch)
    println!("=== Trait Object Usage (Dynamic Dispatch) ===");
    let shapes: Vec<Box<dyn Shape>> = vec![
        Box::new(Circle::new(3.0)),
        Box::new(Rectangle::new(2.0, 4.0)),
        Box::new(Square::new(3.0)),
    ];
    
    for shape in &shapes {
        print_shape_info(shape.as_ref());
        println!();
    }
    
    // Demonstrate generic function (static dispatch)
    println!("=== Generic Function Usage (Static Dispatch) ===");
    compare_shapes(&circle, &rectangle);
    println!();
    
    // Demonstrate ShapeCollection
    println!("=== Shape Collection Demo ===");
    let mut collection = ShapeCollection::new();
    collection.add(Box::new(Circle::new(5.0)));
    collection.add(Box::new(Rectangle::new(4.0, 6.0)));
    collection.add(Box::new(Triangle::new(3.0, 4.0, 5.0).unwrap()));
    collection.add(Box::new(Square::new(4.0)));
    collection.print_all();
    
    // Demonstrate trait methods
    println!("\n=== Additional Trait Methods ===");
    println!("Circle is congruent to another Circle: {}", circle.is_congruent(&Circle::new(5.0)));
    println!("Rectangle is square: {}", rectangle.is_square());
    println!("Triangle is equilateral: {}", triangle.is_equilateral());
    
    // Demonstrate scaling
    println!("\n=== Scaling Shapes ===");
    let mut scalable_circle = Circle::new(2.0);
    println!("Before scaling: {}", scalable_circle);
    scalable_circle.scale(3.0);
    println!("After scaling by 3: {}", scalable_circle);
    
    // Pattern matching with enums (alternative to polymorphism)
    println!("\n=== Enums as Sum Types ===");
    let shapes_enum = vec![
        ShapeEnum::Circle(5.0),
        ShapeEnum::Rectangle(4.0, 6.0),
        ShapeEnum::Triangle(3.0, 4.0, 5.0),
    ];
    
    for shape in &shapes_enum {
        match shape {
            ShapeEnum::Circle(radius) => {
                println!("Circle with radius {}: area = {:.2}", radius, PI * radius * radius);
            }
            ShapeEnum::Rectangle(width, height) => {
                println!("Rectangle {}x{}: area = {:.2}", width, height, width * height);
            }
            ShapeEnum::Triangle(a, b, c) => {
                let s = (a + b + c) / 2.0;
                let area = (s * (s - a) * (s - b) * (s - c)).sqrt();
                println!("Triangle {} {} {}: area = {:.2}", a, b, c, area);
            }
        }
    }
}

// Alternative approach using enums instead of trait objects
#[derive(Debug)]
enum ShapeEnum {
    Circle(f64),
    Rectangle(f64, f64),
    Triangle(f64, f64, f64),
}

impl ShapeEnum {
    fn area(&self) -> f64 {
        match self {
            ShapeEnum::Circle(radius) => PI * radius * radius,
            ShapeEnum::Rectangle(width, height) => width * height,
            ShapeEnum::Triangle(a, b, c) => {
                let s = (a + b + c) / 2.0;
                (s * (s - a) * (s - b) * (s - c)).sqrt()
            }
        }
    }
    
    fn perimeter(&self) -> f64 {
        match self {
            ShapeEnum::Circle(radius) => 2.0 * PI * radius,
            ShapeEnum::Rectangle(width, height) => 2.0 * (width + height),
            ShapeEnum::Triangle(a, b, c) => a + b + c,
        }
    }
}

// Add downcast_ref to Shape trait for concrete type checking
trait ShapeExt: Shape {
    fn downcast_ref<T: 'static>(&self) -> Option<&T> {
        // In a real implementation, you'd use Any trait
        None
    }
}

impl<T: Shape + 'static> ShapeExt for T {}